### PR TITLE
Adds support for Debian

### PR DESCRIPTION
**Pull request by aweidner**
Adds support for Debian operating systems while leaving support for RedHat.  Tested in Ubuntu 16.04.

### CODE REVIEW

**Comment by jtyr** [comment]
Thanks for the PR. Please could you keep the tasks in a single file instead of splitting it per OS? Use `package` instead of `apt` and make the key and repo URL as a variable.

**Comment by aweidner** [comment]
I think this should take care of the concerns in your comment, if there are additional changes needed let me know.  Thanks!

File: defaults/main.yaml [hunk 1]
```diff
@@ -3,6 +3,9 @@
 # Package to install (you can force certain version here)
 rabbitmq_pkg: rabbitmq-server
 
+# APT key for Debian repo
+rabbitmq_debian_aptrepo_key: "https://www.rabbitmq.com/rabbitmq-release-signing-key.asc"
```
Comment (jtyr) [inline defaults/main.yaml]: The value should not be quoted.

File: tasks/main.yaml [hunk 2]
```diff
@@ -1,12 +1,5 @@
 ---
-
-- name: Test if distribution is RedHat
-  assert:
-    that: >
-      ansible_os_family == "RedHat" and
```
Comment (jtyr) [inline tasks/main.yaml]: Please do not remove this. Just extend it to check also for Debian:

```
    that: >
      (
        ansible_os_family == "RedHat" and
        ansible_distribution_major_version | int >= 6
      ) or
      ansible_os_family == "Debian"
```

File: tasks/main.yaml [hunk 3]
```diff
@@ -1,12 +1,5 @@
...
-      ansible_distribution_major_version | int >= 6
-
-- name: Add EPEL YUM repo
+- name: Add EPEL YUM repo (Red Hat)
```
Comment (jtyr) [inline tasks/main.yaml]: Please remove the `(Red Hat)` as it's clear that YUM repo is for RedHat only.

File: tasks/main.yaml [hunk 4]
```diff
@@ -16,14 +9,40 @@
...
+  notify:
+    - Restart rabbitmq-server
+  tags:
+    - rabbitmq_pkg
+  when: ansible_os_family == "RedHat" and
+    ansible_distribution_major_version | int >= 6
+
+- name: Add apt repository
+  apt_repository:
+    repo: deb http://www.rabbitmq.com/debian/ testing main
```
Comment (jtyr) [inline tasks/main.yaml]: This should as a variable.

File: tasks/main.yaml [hunk 5]
```diff
@@ -16,14 +9,40 @@
...
+  apt_key:
+    url: "{{ rabbitmq_debian_aptrepo_key }}"
+    state: "present"
+  when: ansible_os_family == "Debian"
 
-- name: Install rabbitmq-server
+- name: Install rabbitmq-server (Debian)
   package:
     name: "{{ rabbitmq_pkg }}"
+    update_cache: yes
```
Comment (jtyr) [inline tasks/main.yaml]: I kind of don't like to split package installation for RedHat and Debian. Please remove the above task called `Install rabbitmq-server (RedHat)` and remove the `update_cache` and `when` from this task.

File: tasks/main.yaml [hunk 6]
```diff
@@ -33,6 +52,7 @@
     - Restart rabbitmq-server
   tags:
     - rabbitmq_config
+  when: ansible_os_family == "Debian"
```
Comment (jtyr) [inline tasks/main.yaml]: Why is this `when` here?

File: templates/rabbitmq.conf.j2 [hunk 7]
```diff
@@ -1,3 +1,5 @@
 {{ ansible_managed | comment('erlang') }}
 
-{{ rabbitmq_config | encode_erlang(convert_bools=true, convert_nums=true) }}
+[
```
Comment (jtyr) [inline templates/rabbitmq.conf.j2]: Why did you add the square brackets here?

**Review by jtyr — CHANGES_REQUESTED** [review]

File: templates/rabbitmq.conf.j2 (hunk 7 above)
Comment (jtyr) [reply templates/rabbitmq.conf.j2]: I think there is a typo in the examples. The custom config should start like an array as it is in the default config in `defaults/main.yaml`. Then adding the square brackets in the template file is not required.

File: tasks/main.yaml [hunk 8]
```diff
@@ -1,10 +1,12 @@
 ---
-
-- name: Test if distribution is RedHat
+- name: Test if distribution is RedHat or Debian
```
Comment (jtyr) [inline tasks/main.yaml]: Please change the title to "Distribution test"

File: tasks/main.yaml [hunk 9]
```diff
@@ -13,10 +15,25 @@
     baseurl: "{{ rabbitmq_epel_yumrepo_url }}"
     gpgkey: "{{ rabbitmq_epel_yumrepo_gpgkey }}"
     params: "{{ rabbitmq_epel_yumrepo_params }}"
-  when: rabbitmq_epel_install == true
+  when: rabbitmq_epel_install == true and
```
Comment (jtyr) [inline tasks/main.yaml]: Please change this to this:

```
when: >
  rabbitmq_epel_install == true and
  ansible_os_family == "RedHat" and
  ansible_distribution_major_version | int >= 6
```

**Review by jtyr — CHANGES_REQUESTED** [review]
Please could you also fix the typo in the second example of the `README.md` file?

```
# Example of how to create a custom configuration
- hosts: myhost1
  vars:
    # Here goes the custom configuration
    rabbitmq_config:
      - rabbit:
        - tcp_listeners:
          - '"0.0.0.0"': 5672
        - ssl_listeners:
          - 5671
        - ssl_options:
          - cacertfile: /path/to/testca/cacert.pem
          - certfile: /path/to/server/cert.pem
          - keyfile: /path/to/server/key.pem
          - verify: verify_peer
          - fail_if_no_peer_cert: true
  roles:
    - rabbitmq
```

File: README.md [hunk 10]
```diff
@@ -30,7 +30,7 @@ Examples
     - role: rabbitmq
       # Here goes the custom configuration
       rabbitmq_config:
-        rabbit:
+        - rabbit:
```
Comment (jtyr) [inline README.md]: Please replace the whole example with this:

```
# Example of how to create a custom configuration
- hosts: myhost1
  vars:
    # Here goes the custom configuration
    rabbitmq_config:
      - rabbit:
        - tcp_listeners:
          - '"0.0.0.0"': 5672
        - ssl_listeners:
          - 5671
        - ssl_options:
          - cacertfile: /path/to/testca/cacert.pem
          - certfile: /path/to/server/cert.pem
          - keyfile: /path/to/server/key.pem
          - verify: verify_peer
          - fail_if_no_peer_cert: true
  roles:
    - rabbitmq
```

File: tasks/main.yaml [hunk 11]
```diff
@@ -13,10 +15,26 @@
...
+  apt_key:
+    url: "{{ rabbitmq_debian_aptrepo_key }}"
+    state: "present"
+  when: ansible_os_family == "Debian"
+
+- name: Add apt repository
+  apt_repository:
+    repo: "{{ debian_apt_repository }}" 
+    state: present
+    update_cache: true
```
Comment (jtyr) [inline tasks/main.yaml]: Please remove this line.

File: tasks/main.yaml [hunk 12]
```diff
@@ -13,10 +15,26 @@
...
+    ansible_os_family == "RedHat" and
+    ansible_distribution_major_version | int >= 6
   tags:
     - rabbitmq_pkg
 
+- name: Add apt key
+  apt_key:
+    url: "{{ rabbitmq_debian_aptrepo_key }}"
+    state: "present"
+  when: ansible_os_family == "Debian"
```
Comment (jtyr) [inline tasks/main.yaml]: Let's use the same formatting across the whole file:

```
when: >
  ansible_os_family == "Debian"
```

**Review by jtyr — CHANGES_REQUESTED** [review]

File: tasks/main.yaml (hunk 11 above)
Comment (aweidner) [reply tasks/main.yaml]: I tested after removing the update_cache line and the installation fails without it

File: tasks/main.yaml [hunk 13]
```diff
@@ -13,10 +15,28 @@
...
     params: "{{ rabbitmq_epel_yumrepo_params }}"
-  when: rabbitmq_epel_install == true
+  when: >
+    rabbitmq_epel_install == true and
+    ansible_os_family == "RedHat" and
+    ansible_distribution_major_version | int >= 6
   tags:
     - rabbitmq_pkg
 
+- name: Add apt key
```
Comment (jtyr) [inline tasks/main.yaml]: Please make the APT capital. The same bellow.

File: tasks/main.yaml (hunk 11 above)
Comment (jtyr) [reply tasks/main.yaml]: That's weird as I'm using the `package` module without it in other roles (e.g. https://github.com/jtyr/ansible-kibana).

**Review by jtyr — CHANGES_REQUESTED** [review]

File: tasks/main.yaml (hunk 11 above)
Comment (aweidner) [reply tasks/main.yaml]: Apologies, I must have tested it wrong the first time.  I just tested it again and it's working without the update_cache requirement.

File: tasks/main.yaml [hunk 14]
```diff
@@ -13,10 +15,27 @@
     baseurl: "{{ rabbitmq_epel_yumrepo_url }}"
     gpgkey: "{{ rabbitmq_epel_yumrepo_gpgkey }}"
     params: "{{ rabbitmq_epel_yumrepo_params }}"
-  when: rabbitmq_epel_install == true
+  when: >
+    rabbitmq_epel_install == true and
+    ansible_os_family == "RedHat" and
+    ansible_distribution_major_version | int >= 6
```
Comment (jtyr) [inline tasks/main.yaml]: This can be removed as that's tested by the `assert` above.

File: tasks/main.yaml [hunk 15]
```diff
@@ -13,10 +15,27 @@
...
   tags:
     - rabbitmq_pkg
 
+- name: Add APT key
+  apt_key:
+    url: "{{ rabbitmq_debian_aptrepo_key }}"
+    state: "present"
```
Comment (jtyr) [inline tasks/main.yaml]: This should be not quoted.

File: tasks/main.yaml [hunk 16]
```diff
@@ -1,10 +1,12 @@
 ---
-
-- name: Test if distribution is RedHat
+- name: Distribution test
   assert:
     that: >
-      ansible_os_family == "RedHat" and
-      ansible_distribution_major_version | int >= 6
+      (
+        ansible_os_family == "RedHat" and
```
Comment (jtyr) [inline tasks/main.yaml]: Please could you replace the double quotes by single quotes here and in all `when` clauses bellow?

**Review by jtyr — CHANGES_REQUESTED** [review]

File: tasks/main.yaml [hunk 17]
```diff
@@ -13,10 +15,25 @@
     baseurl: "{{ rabbitmq_epel_yumrepo_url }}"
     gpgkey: "{{ rabbitmq_epel_yumrepo_gpgkey }}"
     params: "{{ rabbitmq_epel_yumrepo_params }}"
-  when: rabbitmq_epel_install == true
+  when: >
+    ansible_os_family == 'RedHat'
```
Comment (jtyr) [inline tasks/main.yaml]: Here you forgot the original variable check. Please change it to this:

```
when: >
  ansible_os_family == 'RedHat' and
  rabbitmq_epel_install == true
```

**Review by jtyr — CHANGES_REQUESTED** [review]

File: tasks/main.yaml (hunk 17 above)
Comment (aweidner) [reply tasks/main.yaml]: Ah whoops, you are correct it's obvious when looking at the diff

**Review by jtyr — APPROVED** [review]

**Comment by jtyr** [comment]
Thanks for your contribution!
